File: test/exec.test.ts
/**
 * Unit tests for the shell-out helper. Uses tiny, universally-present commands
 * (sh/printf/false) so it stays deterministic and offline — no kubectl/gh/git.
 */

import { test, expect, describe } from "bun:test"
import { run, runJson, ExecError, JsonParseError, isExecError } from "../src/exec.ts"

describe("run", () => {
  test("captures stdout and exit code 0", async () => {
    const r = await run("printf", ["hello"])
    expect(r.code).toBe(0)
    expect(r.stdout).toBe("hello")
  })

  test("non-zero exit throws an ExecError with the code", async () => {
    let caught: unknown
    try {
      await run("sh", ["-c", "echo boom 1>&2; exit 3"])
    } catch (e) {
      caught = e
    }
    expect(isExecError(caught)).toBe(true)
    const err = caught as ExecError
    expect(err.code).toBe(3)
    expect(err.stderr).toContain("boom")
    expect(err.timedOut).toBe(false)
  })

  test("throwOnError:false returns the non-zero code instead of throwing", async () => {
    const r = await run("sh", ["-c", "exit 7"], { throwOnError: false })
    expect(r.code).toBe(7)
  })

  test("timeout kills the process and flags timedOut", async () => {
    let caught: unknown
    try {
      await run("sh", ["-c", "sleep 5"], { timeoutMs: 50 })
    } catch (e) {
      caught = e
    }
    expect(isExecError(caught)).toBe(true)
    expect((caught as ExecError).timedOut).toBe(true)
  })

  test("timeout still fires for a SIGTERM-ignoring child (SIGKILL escalation)", async () => {
    // Trap and ignore SIGTERM, then loop. A SIGTERM-only kill would hang the
    // await forever; SIGKILL escalation must still end it. Allow generous time
    // for the 2s grace period (default timeout is 120s, so we won't hit it).
    let caught: unknown
    const started = Date.now()
    try {
      await run("sh", ["-c", "trap '' TERM; while true; do sleep 1; done"], {
        timeoutMs: 50,
      })
    } catch (e) {
      caught = e
    }
    expect(isExecError(caught)).toBe(true)
    expect((caught as ExecError).timedOut).toBe(true)
    // Should resolve within the grace window, not hang indefinitely.
    expect(Date.now() - started).toBeLessThan(10_000)
  }, 15_000)

  test("missing binary → ExecError (no throw on spawn)", async () => {
    let caught: unknown
    try {
      await run("kc-definitely-not-a-real-binary-xyz", [])
    } catch (e) {
      caught = e
    }
    expect(isExecError(caught)).toBe(true)
    expect((caught as ExecError).code).toBeNull()
  })

  test("env override is passed through", async () => {
    const r = await run("sh", ["-c", "printf %s \"$KC_TEST_VAR\""], {
      env: { KC_TEST_VAR: "xyzzy" },
    })
    expect(r.stdout).toBe("xyzzy")
  })
})

describe("runJson", () => {
  test("parses JSON stdout", async () => {
    const out = await runJson<{ a: number; b: string[] }>("printf", ['{"a":1,"b":["x"]}'])
    expect(out).toEqual({ a: 1, b: ["x"] })
  })

  test("invalid JSON → JsonParseError", async () => {
    let caught: unknown
    try {
      await runJson("printf", ["not json"])
    } catch (e) {
      caught = e
    }
    expect(caught).toBeInstanceOf(JsonParseError)
  })
})

File: internal/exec/exec_test.go
package exec

import (
	"context"
	"errors"
	"testing"
	"time"
)

// These tests use tiny, universally-present commands (sh/printf/false) so they
// stay deterministic and offline — no kubectl/gh/git.
// Ported from tools/kc-bun/test/exec.test.ts.

func TestRun_CapturesStdoutAndExitZero(t *testing.T) {
	r, err := Run(context.Background(), "printf", []string{"hello"}, RunOptions{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if r.Code != 0 {
		t.Errorf("code = %d, want 0", r.Code)
	}
	if r.Stdout != "hello" {
		t.Errorf("stdout = %q, want %q", r.Stdout, "hello")
	}
}

func TestRun_NonZeroExitReturnsExecError(t *testing.T) {
	_, err := Run(context.Background(), "sh", []string{"-c", "echo boom 1>&2; exit 3"}, RunOptions{})
	ee, ok := AsExecError(err)
	if !ok {
		t.Fatalf("expected ExecError, got %T: %v", err, err)
	}
	if ee.Code != 3 {
		t.Errorf("code = %d, want 3", ee.Code)
	}
	if !contains(ee.Stderr, "boom") {
		t.Errorf("stderr %q does not contain boom", ee.Stderr)
	}
	if ee.TimedOut {
		t.Error("timedOut = true, want false")
	}
	// The stderr tail should be folded into the message.
	if !contains(ee.Error(), "boom") {
		t.Errorf("error message %q omits stderr tail", ee.Error())
	}
}

func TestRun_AllowNonZeroReturnsCode(t *testing.T) {
	r, err := Run(context.Background(), "sh", []string{"-c", "exit 7"}, RunOptions{AllowNonZero: true})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if r.Code != 7 {
		t.Errorf("code = %d, want 7", r.Code)
	}
}

func TestRun_TimeoutKillsAndFlagsTimedOut(t *testing.T) {
	_, err := Run(context.Background(), "sh", []string{"-c", "sleep 5"}, RunOptions{Timeout: 50 * time.Millisecond})
	ee, ok := AsExecError(err)
	if !ok {
		t.Fatalf("expected ExecError, got %T: %v", err, err)
	}
	if !ee.TimedOut {
		t.Error("timedOut = false, want true")
	}
}

func TestRun_TimeoutEscalatesToSIGKILL(t *testing.T) {
	// Trap and ignore SIGTERM, then loop. A SIGTERM-only kill would hang the
	// wait forever; the WaitDelay SIGKILL escalation must still end it.
	start := time.Now()
	_, err := Run(context.Background(), "sh",
		[]string{"-c", "trap '' TERM; while true; do sleep 1; done"},
		RunOptions{Timeout: 50 * time.Millisecond})
	ee, ok := AsExecError(err)
	if !ok {
		t.Fatalf("expected ExecError, got %T: %v", err, err)
	}
	if !ee.TimedOut {
		t.Error("timedOut = false, want true")
	}
	if elapsed := time.Since(start); elapsed > 10*time.Second {
		t.Errorf("took %v, expected resolution within the grace window", elapsed)
	}
}

func TestRun_MissingBinaryReturnsExecError(t *testing.T) {
	_, err := Run(context.Background(), "kc-definitely-not-a-real-binary-xyz", nil, RunOptions{})
	ee, ok := AsExecError(err)
	if !ok {
		t.Fatalf("expected ExecError, got %T: %v", err, err)
	}
	if ee.Code != -1 {
		t.Errorf("code = %d, want -1 for a launch failure", ee.Code)
	}
}

func TestRun_EnvOverridePassedThrough(t *testing.T) {
	r, err := Run(context.Background(), "sh", []string{"-c", `printf %s "$KC_TEST_VAR"`},
		RunOptions{Env: []string{"KC_TEST_VAR=xyzzy"}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if r.Stdout != "xyzzy" {
		t.Errorf("stdout = %q, want xyzzy", r.Stdout)
	}
}

func TestRun_ParentContextCancellationNotFlaggedAsTimeout(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	cancel() // already cancelled
	_, err := Run(ctx, "sh", []string{"-c", "sleep 5"}, RunOptions{Timeout: 5 * time.Second})
	ee, ok := AsExecError(err)
	if !ok {
		t.Fatalf("expected ExecError, got %T: %v", err, err)
	}
	// Parent cancellation is not our deadline → not flagged as timedOut.
	if ee.TimedOut {
		t.Error("parent cancellation should not be reported as a timeout")
	}
}

func TestRunJSON_ParsesStdout(t *testing.T) {
	var out struct {
		A int      `json:"a"`
		B []string `json:"b"`
	}
	if err := RunJSON(context.Background(), "printf", []string{`{"a":1,"b":["x"]}`}, RunOptions{}, &out); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if out.A != 1 || len(out.B) != 1 || out.B[0] != "x" {
		t.Errorf("decoded = %+v, want {A:1 B:[x]}", out)
	}
}

func TestRunJSON_InvalidJSONReturnsJsonParseError(t *testing.T) {
	var out map[string]any
	err := RunJSON(context.Background(), "printf", []string{"not json"}, RunOptions{}, &out)
	var jpe *JsonParseError
	if !errors.As(err, &jpe) {
		t.Fatalf("expected *JsonParseError, got %T: %v", err, err)
	}
}

func contains(haystack, needle string) bool {
	return len(needle) == 0 || (len(haystack) >= len(needle) && indexOf(haystack, needle) >= 0)
}

func indexOf(s, sub string) int {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return i
		}
	}
	return -1
}
